### PR TITLE
Optionally clean up plugin data on uninstall

### LINKED ISSUE #883: Do a clean uninstall
Once the plugin gets uninstalled, remove all plugin data and options from the DB. This can be done by either using the `register_uninstall_hook` action or creating an `uninstall.php` file (similar to WooCommerce core).

WooCommerce core checks for a constant called `WC_REMOVE_ALL_DATA` and only removes data if it has been defined. We could follow a similar approach to make this optional for the user. This also prevents accidental removals.

Some of the data that needs to be removed includes:

- Custom tables
- Scheduled actions in ActionScheduler
- Post metadata
- Plugin options

### PR DESCRIPTION
**Pull request by nima-karimi**
### Changes proposed in this Pull Request:

Closes #883.

This PR adds an `uninstall.php` file that WordPress runs when the plugin is being uninstalled. We check for a constant called `WC_GLA_REMOVE_ALL_DATA` before removing any data in order to prevent data loss.

The data that gets removed includes:

- Custom tables
- Options
- Transients
- Product metadata

Any pending ActionScheduler jobs are also canceled.

Since we don't have access to any of our classes in the `uninstall.php` file, I tried to mostly use native WordPress functions to do the operations and not call any of our services to reduce the dependency. However, I had to initialize the autoloader in order to have access to a bunch of helper methods and constants that were defined in our classes (table names, options, meta keys, etc.).

The only raw query used here is for dropping database tables. Options, transients, and product metadata are all deleted using WordPress native methods.

I'm a little bit worried about the performance of the `delete_post_meta_by_key` function because it runs at least two queries internally and if the number of products is large, it could slow things down and even lead to timeout for the uninstall process. 

I originally intended to use a `DELETE ... LIKE '_wc_gla_%` raw query to delete the posts' metadata but since that has also its own disadvantages (long-running query, not explicitly defining the meta key to delete), I went with the WordPress native functions.

Also for deleting Options and Transients, instead of running raw delete queries (which would be similar to `DELETE ... LIKE 'gla_%`), I went with getting the list of all keys and then deleting them one by one. This is the only way to be sure that we don't delete anything we shouldn't have.

All ActionScheduler jobs using the `gla` group are also unscheduled by using ActionScheduler's own API (`as_unschedule_all_actions`) if the method exists.

A few refactorings are included here too:

- Introduced a new `TableManager` class that keeps a list of all valid custom tables that we add. This was required because `uninstall.php` needed a static list of tables instead of getting them from the container
- Renamed a possible typo (`PluginUpdateJobs`) in `JobServiceProvider`
- Made the `get_raw_name` of `Table` classes static and public. The raw name of a table never changes during the lifetime of any Table class; therefore it made sense for them to be static
- Moved the Options and Transients keys into their interface classes. The keys were all defined in interfaces, so it made sense to also have the list defined there
- Added `get_all_***_keys` to Options, Transients, and ProductMetaHandler classes to return the list of all valid keys they use

### Detailed test instructions:

(It would be best to test this on a separate test site or somewhere that you don't care about losing the data.)

1. Define the `WC_GLA_REMOVE_ALL_DATA` constant in your wp-config.php file and set it to boolean `true`
2. Uninstall the plugin via `WordPress > Plugins`
3. Check that all plugin data is removed and there are no pending GLA action scheduler jobs

### Changelog entry

> Update - Optionally clean up plugin data on uninstall.

### CODE REVIEW

File: src/DB/Table/BudgetRecommendationTable.php [hunk 1]
```diff
@@ -86,8 +86,7 @@ private function load_initial_data(): void {
 		$chunk_size = 500;
 
 		if ( file_exists( $path ) ) {
-			$csv     = array_map( 'str_getcsv', file( $path ) );
-			$headers = array_shift( $csv );
```
Comment (layoutd) [inline src/DB/Table/BudgetRecommendationTable.php:90]: This line was to remove the column headers and not insert them into the database, removing it inserts an invalid row on install:
image

File: src/Options/TransientsInterface.php [hunk 2]
```diff
@@ -41,4 +45,11 @@ public function set( string $name, $value, int $expiration = 0 ): bool;
 	 * @return bool
 	 */
 	public function delete( string $name ): bool;
+
+	/**
+	 * Returns all available transient keys.
+	 *
+	 * @return array
```
Comment (layoutd) [inline src/Options/TransientsInterface.php:52]: Missing `since` tag (do we do that in `interface`s?)

File: src/DB/TableManager.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+use Automattic\WooCommerce\GoogleListingsAndAds\DB\Table\ShippingTimeTable;
+use Automattic\WooCommerce\GoogleListingsAndAds\Exception\ValidateInterface;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class TableManager
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\DB
+ */
```
Comment (layoutd) [inline src/DB/TableManager.php]: Missing `since` tag

File: uninstall.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	require_once __DIR__ . '/src/Autoloader.php';
+	if ( ! Autoloader::init() ) {
+		return;
+	}
+
+	global $wpdb;
+
+	// un-schedule all ActionScheduler jobs for GLA
+	if ( function_exists( 'as_unschedule_all_actions' ) ) {
+		as_unschedule_all_actions( null, null, 'gla' );
```
Comment (layoutd) [inline uninstall.php]: Should this use `WC_GLA_SLUG`?

File: uninstall.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+	// delete transients
+	foreach ( Transients::get_all_transient_keys() as $transient ) {
+		delete_transient( WC_GLA_SLUG . "_{$transient}" );
+	}
+
+	// delete products metadata
+	foreach ( ProductMetaHandler::get_all_meta_keys() as $meta_key ) {
+		delete_post_meta_by_key( WC_GLA_METAKEY_PREFIX . "_{$meta_key}" );
+	}
```
Comment (layoutd) [inline uninstall.php:65]: This leaves behind the GLA Product Attribute values, like `_wc_gla_brand`,  `_wc_gla_gtin`, etc.

**Review by layoutd — CHANGES_REQUESTED** [review]
Looks pretty good generally, definitely goes slowly (hit my 90s execution time limit a couple times). I left a few changes/suggestions, some are pretty minor, but some are more important, most notably the invalid row inserted into BudgetRecommendations and the GLA Product Attributes not being removed.

> It would be best to test this on a separate test site or somewhere that you don't care about losing the data.

I started testing on a test site, then decided to be "smart" and use my main GLA dev site and just copy the DB, but I forgot to copy my IDE settings so now I'm setting it up from 0 again. I'll take another pass at it tonight/tomorrow, but I wanted to get these comments out first.

File: src/DB/Table/BudgetRecommendationTable.php (hunk 1 above)
Comment (nima-karimi) [reply src/DB/Table/BudgetRecommendationTable.php:90]: Ops I thought it was leftover from a merge conflict. Reverted it and added a comment there so the future me doesn't delete it again 🤦‍♂️

File: uninstall.php (hunk 4 above)
Comment (nima-karimi) [reply uninstall.php]: added in https://github.com/woocommerce/google-listings-and-ads/pull/892/commits/b48f067762b6df32c746d9bb79e37f657808b0b4

File: uninstall.php (hunk 5 above)
Comment (nima-karimi) [reply uninstall.php:65]: Oh yeah, I totally missed those. Thanks for catching it. Added the clean up code for product attributes in https://github.com/woocommerce/google-listings-and-ads/pull/892/commits/8d263e8fa63377f79c0b336af02eb78594cb13c3

**Comment by nima-karimi** [comment]
Thanks for the review @layoutd 🙏 This is ready for another go.

> definitely goes slowly (hit my 90s execution time limit a couple times)

Yeah, it could get incredibly slow if there are tons of products and data. I don't think we can do much about it though except asking the user to increase their timeout limit. Using raw SQL might reduce the execution time to some degree but it would still be pretty slow.

**Review by layoutd — APPROVED** [review]
LGTM 💯 

I ran a few tests with a built package and it was noticeably faster (not deleting `node_modules` or `vendor` w/dev deps ... and maybe `.git` in my last tests 🙈 ).

**Pull request merged**